### ISSUE TITLE
Use ALPN "istio" for PERMISSIVE mode.

### ISSUE THREAD

**Issue by incfly** [istio/istio#6384]
Current implementation relies on `envoy.tls_inspector`. @PiotrSikora mentions now we ship ALPN in release 0.8. The outcome to authn policy is effectively the same, telling whether it's a traffic originating from istio sidecar or not.

Questions, what does each filter chain match looks like for istio TLS and plaintext respectively?

**Comment by PiotrSikora** [comment]
```
listeners:
- address:
    socket_address: { address: 127.0.0.1, port_value: 10000 }
  listener_filters:
  - name: "envoy.listener.tls_inspector"
    config: {}
  filter_chains:
  - filter_chain_match:
      application_protocols: "istio"
    tls_context:
      common_tls_context:
        tls_certificates:
        - certificate_chain: { filename: "test/common/ssl/test_data/san_dns_cert.pem" }
          private_key: { filename: "test/common/ssl/test_data/san_dns_key.pem" }
    filters:
    - name: ...
      [...]
  - filters:
    - name: ...
       [...]
```

**Comment by incfly** [comment]
Do you still need "envoy.listener.tls_inspector" to use ALPN?

**Comment by PiotrSikora** [comment]
Yes.

**Comment by incfly** [comment]
So the inspector look at FilterChainMatch's contained fields, and may do TLS sniffing or not depends on whether `filter_chain_match.transport_protocol` is set?

I thought one of the ALPN's advantages is to avoid sniffing.

**Comment by incfly** [comment]
One more thing to confirm, so with [above LDS config](https://github.com/istio/istio/issues/6384#issuecomment-398251823), both workload-to-workload TLS and plain text traffic will be dispatched to the second filter chain. In that case, it's application itself to terminate TLS  or simply handle plaintext traffic.

In short, the filter chain without any FilterChainMatch field specified acts as a `default` one.

Is that right?

**Comment by PiotrSikora** [comment]
> One more thing to confirm, so with above LDS config, both workload-to-workload TLS and plain text traffic will be dispatched to the second filter chain. In that case, it's application itself to terminate TLS or simply handle plaintext traffic.

Correct. But, if you also want to terminate TLS in the sidecar, then you can simply add another filter chain (3 in total) to allow mTLS, "legacy" TLS terminated at the sidecar, and unencrypted traffic. Using the same example:

```
listeners:
- address:
    socket_address: { address: 127.0.0.1, port_value: 10000 }
  listener_filters:
  - name: "envoy.listener.tls_inspector"
    config: {}
  filter_chains:
  - filter_chain_match:
      transport_protocol: "tls"
      application_protocols: "istio"
    tls_context:
      common_tls_context:
        tls_certificates:
        - certificate_chain: { filename: "test/common/ssl/test_data/san_dns_cert.pem" }
          private_key: { filename: "test/common/ssl/test_data/san_dns_key.pem" }
    filters:
    - name: ...
      [...]
  - filter_chain_match:
      transport_protocol: "tls"
    tls_context:
      common_tls_context:
        tls_certificates:
        - certificate_chain: { filename: "test/common/ssl/test_data/san_dns_cert.pem" }
          private_key: { filename: "test/common/ssl/test_data/san_dns_key.pem" }
    filters:
    - name: ...
      [...]
  - filters:
    - name: ...
      [...]
```

> In short, the filter chain without any FilterChainMatch field specified acts as a default one.

In this particular case, yes, but that's not always the case, since match on one of the selectors can limit available choices (I'm working on the documentation for that).

**Comment by andraxylia** [comment]
@incfly please use istio-users for questions, do not open issues unless there is an issue.

**Comment by PiotrSikora** [comment]
@andraxylia This issue tracks switching from TLS multiplexing to `"istio"` ALPN for incremental mTLS, and it's definitely not a question for istio-users@.

**Comment by andraxylia** [comment]
ok. The description sounded like a question.

**Comment by PiotrSikora** [comment]
@andraxylia this should be targeting 1.0, otherwise we're going to ship the half-working solution based on TLS multiplexing (https://github.com/istio/istio/issues/5059).

cc @incfly @diemtvu @wattli @wenchenglu

**Comment by andraxylia** [comment]
@PiotrSikora great if it makes it in 1.0, but according to @louiscryan it is not a P0 for 1.0, he assigned it to 1.1.

**Comment by andraxylia** [comment]
not stale

**Comment by costinm** [comment]
My understanding:
Use case: 
- in permissive mode ( incremental adoption )
- non istio TLS client, without sidecar
- app handles TLS in its own way
- envoy will see a TLS connection - without "istio" alpn - and instead of terminating TLS will forward 
as TCP proxy (plain text from envoy perspective ).

That means envoy acting as client will need to indicate "istio" in ALPN, and server will need to do the same so they negotiate it. The tricky part is in envoy server side so "h2" can be negotiated as well.

It would be good to find a different word than istio - and maybe figure if IETF could clarify this kind of use.

**Comment by incfly** [comment]
Fixed via https://github.com/istio/istio/pull/6612

**Comment by incfly** [comment]
Reopen the issue for adding an e2e test for https service behind `PERMISSIVE` mode.

We recently hit some issue that case does not work and magically work again in 1.1 branch and master... 
https://github.com/istio/istio/issues/9837

**Comment by incfly** [comment]
#10062 is another potential victim...

**Comment by duderino** [comment]
@PiotrSikora @incfly which one of you is working on this?

**Comment by incfly** [comment]
Let's close this for 1.1 since we now have test for Envoy configuration. I'll file a separate issue https://github.com/istio/istio/issues/11566
